### ISSUE TITLE
Lexer: Chaining: Leading dot throws error when previous line ends with empty function

### ISSUE THREAD

**Issue by bioball** [jashkenas/coffeescript#3845]
This is not a huge deal, but not the behavior I was expecting. I'd like to be able to chain method calls with function callbacks together, even if the function bodies were empty. The use case is for cases like this:

``` coffeescript
doThing()
.then (@result) =>
.catch handleError
```

Where I'd expect this to compile into:

``` js
doThing().then((function(_this) {
  return function(account) {
    _this.account = account;
  };
})(this))["catch"](handleError);
```

Currently, this just throws an error: `Error on line 3: unexpected .`

**Comment by edgebal** [comment]
@bioball you can use parentheses to encapsulate empty functions.

```
oThing()
.then ( (@result) => )
.catch handleError
```

**Comment by bioball** [comment]
Yup, that's what I'm currently doing. It's not such much that I can't do it, it's just that this works:

```coffeescript
doThing()
.then (@result) => undefined
.catch handleError
```

While this doesn't

```coffeescript
doThing()
.then (@result) =>
.catch handleError
```

Adding parentheses is fine, I just expected functions with empty bodies to work the same way functions with bodies do.

**Comment by epidemian** [comment]
> I just expected functions with empty bodies to work the same way functions with bodies do.

I would expect the same. Especially considering that:

```coffeescript
foo ->
bar ->
```

... compiles to:

```javascript
foo(function() {});
bar(function() {});
```

... and:

```coffeescript
obj
.foo ->
```

... compiles to:

```javascript
obj.foo(function() {});
```

Then:

```coffeescript
obj
.foo ->
.bar ->
```

... should compile to:

```javascript
obj.foo(function() {}).bar(function() {});
```

Unless i'm missing something, this seems like a bug.

**Comment by nfour** [comment]
@epidemian 

You mean:

```coffee
obj
.foo ->
.bar ->
```
Should compile to

```coffee
_ref = obj
_ref.foo(function() {})
_ref.bar(function() {})
```

**Comment by Panoplos** [comment]
I am having a similar issue with *bluebird* `Promise` code, as follows:

```coffeescript
doSomething argument
  .then (res) ->
    console.log res
  .catch (err) ->
    console.error err
```

this result in the following JavaScript:

```javascript
doSomething(argument).then(function(res) {
  console.log(res);
})["catch"](function(err) {
  console.error(err);
});
```

Which of course is seriously wrong...

**Comment by nfour** [comment]
@Panoplos `catch` is a reserved word, it's for ES4 compat or something like that.
That code, however, is compiling correctly is it not?

**Comment by michaelficarra** [comment]
@nfour: ES3.

**Comment by GeoffreyBooth** [comment]
Fixed by #4588.
